### PR TITLE
[Speculative Decoding] Refactor Eagle MTP hidden states copy

### PR DESCRIPTION
**Pull request by huicongyao**
## Motivation

> :bulb: If this PR is a Cherry Pick, the PR title needs to follow the format by adding the [Cherry-Pick] label at the very beginning and appending the original PR ID at the end. For example, [Cherry-Pick][CI] Add check trigger and logic(#5191)

> :bulb: 如若此PR是Cherry Pick，PR标题需遵循格式，在最开始加上[Cherry-Pick]标签，以及最后面加上原PR ID，例如[Cherry-Pick][CI] Add check trigger and logic(#5191)

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

**Review by Sunny-bot1 — DISMISSED** [review]

File: fastdeploy/spec_decode/mtp.py [hunk 1]
```diff
@@ -127,8 +127,11 @@ def __init__(
         self.model_inputs.init_share_inputs()
 
         if current_platform.is_cuda() or current_platform.is_maca():
-            self._real_output_token_num_host = paddle.empty([1], dtype="int32").pin_memory()
-            self.output_token_num_event = paddle.device.cuda.Event()
+            self._mtp_input_token_num_host = paddle.empty([1], dtype="int32").pin_memory()
+            self._draft_output_token_num_host = paddle.empty([1], dtype="int32").pin_memory()
```
Comment (freeliuzc) [inline fastdeploy/spec_decode/mtp.py]: mtp input 和 draft_output 名字不匹配，需要修正下

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 2]
```diff
@@ -1291,34 +1291,13 @@ def forward_cuda(
...
-            sampling_metadata.step_idx,
-            sampling_metadata.min_dec_lens,
-            sampling_metadata.eos_token_ids,
-            share_inputs["seq_lens_this_time"],
-            share_inputs["batch_id_per_token_output"],
-            share_inputs["cu_seqlens_q_output"],
-            max_model_len,
-            sampling_metadata.pre_token_ids,
-        )
-        probs = F.softmax(logits)
```
Comment (Deleter-D) [inline fastdeploy/model_executor/layers/sample/sampler.py:1319]: 这些逻辑去掉的原因是什么

Comment (huicongyao) [reply fastdeploy/model_executor/layers/sample/sampler.py:1319]: draft token的采样应该不需要做penalty和softmax？

File: fastdeploy/worker/input_batch.py [hunk 3]
```diff
@@ -978,6 +978,7 @@ def reset_model_inputs(self) -> None:
             # Reset last sequence lengths if applicable
             if self.num_model_steps > 1:
                 fill_paddle_tensor(self, "last_seq_lens_this_time", -1)
+                fill_paddle_tensor(self, "last_seq_lens_encoder", 0)
```
Comment (freeliuzc) [inline fastdeploy/worker/input_batch.py]: 这里应该不需要了？

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 4]
```diff
@@ -1290,35 +1290,7 @@ def forward_cuda(
             elif self.logprobs_mode == "raw_logits":
                 raw_logprobs = share_inputs["draft_logits"][:real_token_num, :].clone()
 
-        if sampling_metadata.token_ids_all is not None:
-            token_ids_all = sampling_metadata.token_ids_all
-            prompt_lens = sampling_metadata.prompt_lens
-        else:
-            token_ids_all = sampling_metadata.pre_token_ids
```
Comment (freeliuzc) [inline fastdeploy/model_executor/layers/sample/sampler.py:1297]: 前面这里的 token_ids_all / prompt_lens 不能去掉

File: custom_ops/gpu_ops/speculate_decoding/draft_model/eagle_get_hidden_states.cu [hunk 5]
```diff
@@ -141,60 +152,92 @@ std::vector<paddle::Tensor> DispatchDtype(
...
-
-  int output_token_num_cpu =
-      output_token_num.copy_to(paddle::CPUPlace(), false).data<int>()[0];
-
-  auto out = paddle::empty(
-      {output_token_num_cpu, dim_embed}, input.dtype(), input.place());
+                  input_token_num * sizeof(int),
+                  input.stream());
+
+  auto output_token_num =
```
Comment (freeliuzc) [inline custom_ops/gpu_ops/speculate_decoding/draft_model/eagle_get_hidden_states.cu:165]: 这个还需要吗？

**Review by freeliuzc — APPROVED** [review]

**Review by yuanlehome — APPROVED** [review]
